src/Files.App/Actions/FileSystem/Transfer/BaseTransferItemAction.cs

```diff
@@ -0,0 +1,129 @@
+		protected readonly StatusCenterViewModel StatusCenterViewModel = Ioc.Default.GetRequiredService<StatusCenterViewModel>();
+
+		public bool IsExecutable
+			=> ContentPageContext.HasSelection;
+
+		public BaseTransferItemAction()
+		{
+			ContentPageContext.PropertyChanged += ContentPageContext_PropertyChanged;
+		}
+
+		public async Task ExecuteTransferAsync(DataPackageOperation type = DataPackageOperation.Copy)
+		{
+			if (ContentPageContext.ShellPage is null ||
+				ContentPageContext.ShellPage.SlimContentPage.IsItemSelected is false)
+				return;
+
+			// Reset cut mode
+			ContentPageContext.ShellPage.SlimContentPage.ItemManipulationModel.RefreshItemsOpacity();
+
+			ConcurrentBag<IStorageItem> items = [];
+			var itemsCount = ContentPageContext.SelectedItems.Count;
+			var statusCenterItem = itemsCount > 50 ? StatusCenterHelper.AddCard_Prepare() : null;
+			var dataPackage = new DataPackage() { RequestedOperation = type };
+
+			try
+			{
+				// Update the status to in-progress
+				if (statusCenterItem is not null)
+				{
+					statusCenterItem.Progress.EnumerationCompleted = true;
+					statusCenterItem.Progress.ItemsCount = items.Count;
+					statusCenterItem.Progress.ReportStatus(FileSystemStatusCode.InProgress);
+				}
+
+				await ContentPageContext.SelectedItems.ToList().ParallelForEachAsync(async listedItem =>
+				{
+					// Update the status to increase processed count by one
+					if (statusCenterItem is not null)
+					{
+						statusCenterItem.Progress.AddProcessedItemsCount(1);
+						statusCenterItem.Progress.Report();
+					}
+
+					if (listedItem is FtpItem ftpItem)
+					{
+						// Don't dim selected items here since FTP doesn't support cut
+						if (ftpItem.PrimaryItemAttribute is StorageItemTypes.File or StorageItemTypes.Folder)
+							items.Add(await ftpItem.ToStorageItem());
+					}
+					else
+					{
+						if (type is DataPackageOperation.Move)
+						{
+							// Dim opacities accordingly
+							await MainWindow.Instance.DispatcherQueue.EnqueueOrInvokeAsync(() =>
+							{
+								listedItem.Opacity = Constants.UI.DimItemOpacity;
+							});
+						}
+
+						FilesystemResult? result =
+							listedItem.PrimaryItemAttribute == StorageItemTypes.File || listedItem is ZipItem
+								? await ContentPageContext.ShellPage.ShellViewModel.GetFileFromPathAsync(listedItem.ItemPath).OnSuccess(t => items.Add(t))
+								: await ContentPageContext.ShellPage.ShellViewModel.GetFolderFromPathAsync(listedItem.ItemPath).OnSuccess(t => items.Add(t));
+
+						if (!result)
+							throw new IOException($"Failed to process {listedItem.ItemPath} in cutting/copying to the clipboard.", (int)result.ErrorCode);
+					}
+				},
+				10,
+				statusCenterItem?.CancellationToken ?? default);
+
+				var standardObjectsOnly = items.All(x => x is StorageFile or StorageFolder or SystemStorageFile or SystemStorageFolder);
```

Comment:
Just a note: using StorageFile and/or StorageFolder for `DataPackage.SetStorageItems` method is fair enough since what this way produces is the exactly same as what File Explorer produces. However, as we’re in a move of avoiding WinRT APIs in an aspect of slowness, I may want to create our own service using Win32 API or store path strings instead of objects of WinRT. With that’s said we could create converter to WinRT objects in order to avoid such hard work.